Feature/fix ack and retransmit

everHannes opened the pull request:
to me it looks to work stable now, and fixes network flooding, packet loss, and abrupt retransmit,
and in part fixes sync to death and aborted transfer issues we experienced

everHannes committed:
fix ack & resend all not already ack'ed data on retransmit

everHannes committed:
fix & ack

everHannes committed:
Merge branch 'master' into feature/fix_ack_and_retransmit

Conflicts:
	utcp.c

everHannes committed:
fix merge retransmit

everHannes committed:
advance snd.nxt also for unacceptable packets to allow recover with no retransmit

everHannes committed:
fix update timers

everHannes committed:
fix triple ack

everHannes committed:
fix connection reset on retransmit

everHannes committed:
speed up utcp print_packet debug output and print in hex

everHannes committed:
speed up utcp print_packet debug output and print in hex

everHannes committed:
fix to update connection timeout in utcp_timeout

everHannes committed:
fix to update rtrx timer on all progress and connection timer on accepted packets

there might be lots of data queued or filled over time, so the rtrx timer also needs to update on progress

everHannes committed:
double retransmit timeout and reset on ack responses to leave the sender more time to process queued incoming packets

everHannes committed:
limit packets in transit to avoid flooding

everHannes committed:
fix to reset congestion window and use hard limit

everHannes committed:
handle overlapping packets

# Conflicts:
#	utcp.c

everHannes committed:
handle overlapping packets

everHannes committed:
increase max pkt hard limit

everHannes committed:
fix len < 0 type conversion

everHannes committed:
fix type conversion

Commit:
Show total bytes sent/received in test program.

Commit:
When debugging, set initial segment seqno to 0.

Commit:
Fix check for return value of malloc().

Commit:
Allow putting data in a buffer at an arbitrary offset.

Commit:
Fix check for return value of malloc(), allow initially zero length buffer.

Commit:
Add a receive buffer.

The receive buffer kicks in the moment we get a packet which is out of order.
We store the packet in the buffer, and keep track of up to 4 ranges of bytes
of received data. When retransmission fills the first gap, we send all the
buffered data (up to the second gap if applicable) to the application.

4 byte ranges seems to be a good value for up to moderate (20%) packet loss.
This algorithm greatly reduces the amount of useless packets being sent.
A future improvement is sending the SACK information in the ACK packets,
so the congestion window can be kept large while avoiding packets being
resent unnecessarily.

Commit:
Fix a memory leak on invalid invocation of utcp_init().

Commit:
Document difference between snd.nxt and snd.last.

Commit:
Improve documentation of retransmission timeout handling.

Commit:
Minor changes in the test program.

- Add one to the timeout in milliseconds, to prevent roundoff errors from
  causing us to busy-loop unnecessarily.
- Log what we are polling for and what the timeout is.

Commit:
Remove accidentily left #warning.

Commit:
Turn magic numbers into #defines.

Commit:
Measure RTT and calculate RTO.

Opportunistically measure RTT using only a single timer, without requiring
timestamps to be added to packets. Use the method described in RFC 6298 to
smoothly update the value of RTO.

Commit:
Use the calculated RTO value to set the retransmission timer.

This significantly cuts down the time it takes to transfer data when there
is packet loss, because retransmission kicks in much earlier.

Commit:
Fix the logic for determining whether a packets has an acceptable ack seqno.

snd.last accurately tracks the last possible seqno that can be acked, so
use it. This fixes a case where retransmitted ACKs of a FIN were not handled
correctly.

Commit:
Get rid of an unused label.

Commit:
Log when we start retransmission.

Commit:
Send ACKs also when receiving out-of-order data.

This is necessary to trigger the triplicate ACK detection on the sender.

vollkorn1982 commented:
@gsliepen I think you should have a look at this pull request.

gsliepen commented:
On Mon, Oct 19, 2015 at 12:28:11AM -0700, vollkorn1982 wrote:

> @gsliepen I think you should have a look at this pull request.

Yes. Part of the fixes from feature/fix_ack_and_retransmit are also in
the feature/rtt_measurement branch in some way or another. I have to
check the timer stuff, and I'll merge the overlapping packet handling
and changes to the seqno checks, I'll have to see about some of the
other commits since some of them are obsoleted by the receive buffer.
I'll do this today.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

Commit:
Make sure all of the packet header is set when retransmitting packets.

Commit:
Fix a memory leak.

The receive buffer was never freed.

Commit:
Simulate packet reordering.

With a configurable chance, delay a packet for a random amount of packets.

Commit:
Allow setting $DROPFROM without having to specify $DROPTO as well.

Commit:
Ensure FIN packets start the retransmission timer as well.

Commit:
Fix bug in retransmit().

The logic to check whether we actually have something to retransmit was
wrong, causing retransmit() to bail out early without setting or resetting
the timer. This also caused utcp_timeout() to return a negative value.

Commit:
Fix bugs in sack_consume() causing data corruption or abort()s.

Put in a lengthy comment with some ASCII-art to describe what we are
actually trying to do in this function.

everHannes committed:
Merge branch 'feature/receive-buffer' into feature/fix_ack_and_retransmit

Conflicts:
	utcp.c

everHannes committed:
fix sack_consume, stop abort on offset+len > rcvbuf.maxsize

everHannes committed:
Merge branch 'feature/rtt-measurement' into feature/fix_ack_and_retransmit

Conflicts:
	utcp.c

everHannes committed:
fix connection timeout

everHannes committed:
clean packet acceptance

everHannes committed:
fix set_state connection_timer typo

everHannes committed:
fix never use %zu!! it's unsupported by windows

everHannes committed:
fix buffer_put_at abort and add some debug output

when on send the return value is used to check whether data could be added to the buffer, don't abort!

a-krebs commented:
I can't find anything obvious wrong with this. @vollkorn1982, can you look this over and then merge?

vollkorn1982 commented on test.c:
```diff
@@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
+			total_out += reorder_len;
+			send(s, reorder_data, reorder_len, MSG_DONTWAIT);
+			free(reorder_data);
+			reorder_data = NULL;
```
Erm, maybe I'm missing something here, but `reorder_data` is not initialized on line 33, so it may be some undefined pointer. Probably a null pointer when built in debug mode and pretty sure random value when built as a release built. Then on line 73 the if would fail but on line 82 the if would succeed and maybe send out a random amount of random memory.
Fix: initialize `reorder_data` properly.

gsliepen replied:
On Thu, Oct 29, 2015 at 03:37:42AM -0700, vollkorn1982 wrote:

> Erm, maybe I'm missing something here, but `reorder_data` is not initialized on line 33, so it may be some undefined pointer. Probably a null pointer when built in debug mode and pretty sure random value when built as a release built. Then on line 73 the if would fail but on line 82 the if would succeed and maybe send out a random amount of random memory.
> Fix: initialize `reorder_data` properly.

It's a global variable, so it is initialized to zero by default. This is
guaranteed by the C standard.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 replied:
Ah, good, didn't realize it's a global.

vollkorn1982 commented on test.c:
```diff
@@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
+		if(drand48() < dropout)
+			return len;
+		if(!reorder_data && drand48() < reorder) {
+			reorder_data = malloc(len);
```
Can you be sure `len` will never be 0? Because with `len = 0` the returned pointer is implementation defined.

gsliepen replied:
On Thu, Oct 29, 2015 at 03:48:00AM -0700, vollkorn1982 wrote:

> > @@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
> >  ssize_t do_send(struct utcp *utcp, const void *data, size_t len) {
> >     int s = *(int *)utcp->priv;
> >     outpktno++;
> > -   if(outpktno < dropto && outpktno >= dropfrom && drand48() < dropout)
> > -       return len;
> > -   if(outpktno >= dropfrom && outpktno < dropto) {
> > -       if(drand48() < dropout)
> > -           return len;
> > -       if(!reorder_data && drand48() < reorder) {
> > -           reorder_data = malloc(len);
> 
> Can you be sure `len` will never be 0? Because with `len = 0` the returned pointer is implementation defined.

Yes, UTCP will never send zero length packets. But the return value from
malloc() should be checked here in any case, even if len > 0.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 commented on test.c:
```diff
@@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
+		if(!reorder_data && drand48() < reorder) {
+			reorder_data = malloc(len);
+			reorder_len = len;
+			memcpy(reorder_data, data, len);
```
And here with a random pointer value in `reorder_data` and `len` being a possible 0 we encounter undefined behaviour, although I'm pretty sure any implementation of memcpy just would do nothing. Still feeling uneasy about this, though, and would like to make this code safer.

vollkorn1982 commented on test.c:
```diff
@@ -82,6 +109,11 @@ int main(int argc, char *argv[]) {
 	dropfrom = atoi(getenv("DROPFROM") ?: "0");
 	dropto = atoi(getenv("DROPTO") ?: "0");
+	reorder = atof(getenv("REORDER") ?: "0");
+	reorder_dist = atoi(getenv("REORDER_DIST") ?: "10");
```
Omitting the ternary operator's first parameter is a GNU extension and should be avoided for portable code. No clue how MSVC would handles this. This probably only works because right now we still compile meshlink with mingw, which we'll try to change to MSVC in the future.

gsliepen replied:
On Thu, Oct 29, 2015 at 03:56:54AM -0700, vollkorn1982 wrote:

> > @@ -82,6 +109,11 @@ int main(int argc, char *argv[]) {
> >     dropout = atof(getenv("DROPOUT") ?: "0");
> >     dropfrom = atoi(getenv("DROPFROM") ?: "0");
> >     dropto = atoi(getenv("DROPTO") ?: "0");
> > -   reorder = atof(getenv("REORDER") ?: "0");
> > -   reorder_dist = atoi(getenv("REORDER_DIST") ?: "10");
> 
> Omitting the ternary operator's first parameter is a GNU extension and should be avoided for portable code. No clue how MSVC would handles this. This probably only works because right now we still compile meshlink with mingw, which we'll try to change to MSVC in the future.

You're right, I'll fix that.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 commented on utcp.c:
```diff
@@ -50,6 +51,10 @@
 #endif
 
+#ifndef max
+#define max(a, b) ((a) > (b) ? (a) : (b))
```
This would call operations twice. For example this would have unexpected results: `max( i++, 10 )` because `i++` would be called twice. It's much safer and no problem performance-wice to make this a small inline function.

vollkorn1982 commented on utcp.c:
```diff
@@ -79,11 +84,24 @@ static void print_packet(struct utcp *utcp, const char *dir, const void *pkt, si
+			++strptr;
+			*strptr = (*data & 0xf) > 9? (*data & 0xf) + 55 : (*data & 0xf) + 48;
+			++strptr;
+			++data;
```
This whole thing seems like magic to me. Why 48? 55? >> 4? A few comments on what these constants mean and this code tries to achieve would be great.

everHannes replied:
I overworked it a bit and added the binToHex function
48 is simply the '0' asci character and 10 + 55 is asci 'A' character
`>> 4` shifts bits and `& 0xf` masks bits to match 0-15 for upper and lower hex letter each byte
it doesn't regard string localization but is a speedy debug output - a huge difference to printing letter by letter

vollkorn1982 commented on utcp.c:
```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+ * - the SACK entry is ahead of ^, in that case just change the offset.
+ * - the SACK entry starts before and ends after ^, so we have to
+ *   change both its offset and size.
+ * - the SACK entry is completely before ^, in that case delete it.
```
I'm confused by `ahead` and `before` in this context. I'd think both would mean the same direction in the buffer, i.e. left of ^.

gsliepen replied:
On Thu, Oct 29, 2015 at 05:56:09AM -0700, vollkorn1982 wrote:

> > - \* |.....0000..1111111111.....22222......3333|
> > - \* |---------------^
> > - *
> > - \* There are three cases:
> > - \* - the SACK entry is ahead of ^, in that case just change the offset.
> > - \* - the SACK entry starts before and ends after ^, so we have to
> > - \*   change both its offset and size.
> > - \* - the SACK entry is completely before ^, in that case delete it.
> 
> I'm confused by `ahead` and `before` in this context. I'd think both would mean the same direction in the buffer, i.e. left of ^.

s/ahead/after/, and after being right of ^.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 commented on utcp.c:
```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+				// move remaining sacks one ahead
+				memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) * sizeof c->sacks[i]);
+				// clean last sack len
+				c->sacks[NSACKS - 1].len = 0;
```
I'm missing a `break` in this case. How are we ending the for loop here?

everHannes replied:
I got confused on this as well, but the sacks are moved left and last sack len is set 0 so by moving the sacks at some point the for loop stops on len = 0 ;-)

vollkorn1982 replied:
Hm. Let's say `NSACKS = 10` and `i = 5`. Now We end up in this branch, move all the sacks to the left, `i` is still 5 and `c->sacks[NSACKS - 1]` resolves to `c->sacks[9].len = 0`. Then we loop again and the for condition is checked: `c->sacks[5].len` did change to the len of the sack which was previously one to the right. We go into this branch again and again move the same amount of sacks one to the left, again copying the last one, making sure its len is set to 0 and thus propagating the len = 0 to the left until it hits position 5. All the sacks [5, NSACKS[ are effectively removed. Isn't there an easier, more straight forward way to just toss out the remaining sacks?

gsliepen replied:
On Thu, Oct 29, 2015 at 08:02:30AM -0700, vollkorn1982 wrote:

> > -   for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
> > -       if(len < c->sacks[i].offset) {
> > -           c->sacks[i].offset -= len;
> > -           i++;
> > -       } else if(len < c->sacks[i].offset + c->sacks[i].len) {
> > -           c->sacks[i].len -= len - c->sacks[i].offset;
> > -           c->sacks[i].offset = 0;
> > -           i++;
> > -       } else {
> > -           if(i < NSACKS - 1) {
> > -               // move remaining sacks one ahead
> > -               memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) \* sizeof c->sacks[i]);
> > -               // clean last sack len
> > -               c->sacks[NSACKS - 1].len = 0;
> 
> Hm. Let's say `NSACKS = 10` and `i = 5`. Now We end up in this branch, move all the sacks to the left, `i` is still 5 and `c->sacks[NSACKS - 1]` resolves to `c->sacks[9].len = 0`. Then we loop again and the for condition is checked: `c->sacks[5].len` did change to the len of the sack which was previously one to the right. We go into this branch again and again move the same amount of sacks one to the left, again copying the last one, making sure its len is set to 0 and thus propagating the len = 0 to the left until it hits position 5. All the sacks [5, NSACKS[ are effectively removed. Isn't there an easier, more straight forward way to just toss out the remaining sacks?

I set NSACKS to 4 after testing how many sack regions were created when
having 20% packet loss. At this level, because of the triplicate ACK
mechanism, the sender will almost certainly start retransmission before
the receiver would have to use more SACK regions. So I wouldn't
increase NSACKS at all. With only 4 sack entries, I don't think you'd
gain anything by adding logic to detect if you can optimize that last
branch.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 replied:
I thought it might be just much easier to understand if this branch did something like

```
for ( int j = i; j < NSACKS; ++j )
    c->sacks[j].len = 0;
break;
```

Wouldn't that have the same outcome?

gsliepen replied:
On Thu, Oct 29, 2015 at 08:22:30AM -0700, vollkorn1982 wrote:

> > -   for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
> > -       if(len < c->sacks[i].offset) {
> > -           c->sacks[i].offset -= len;
> > -           i++;
> > -       } else if(len < c->sacks[i].offset + c->sacks[i].len) {
> > -           c->sacks[i].len -= len - c->sacks[i].offset;
> > -           c->sacks[i].offset = 0;
> > -           i++;
> > -       } else {
> > -           if(i < NSACKS - 1) {
> > -               // move remaining sacks one ahead
> > -               memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) \* sizeof c->sacks[i]);
> > -               // clean last sack len
> > -               c->sacks[NSACKS - 1].len = 0;
> 
> I thought it might be just much easier to understand if this branch did something like
> 
> ```
> for ( int j = i; j < NSACKS; ++j )
>     c->sacks[j].len = 0;
> break;
> ```
> 
> Wouldn't that have the same outcome?

Oh, no. To use the diagram:

```
|.....0000..1111111111.....22222......3333|
|---------------^
```

SACK entry 0 is completely before the ^. So that's when the last branch
comes into play. We don't want to erase all the SACK entries here. What
it does is:

```
|...........0000000000.....11111......2222|
|---------------^
```

So it erased the first SACK entry, and merely shifted the index of the
others. Because the last branch doesn't do i++, the next iteration will
compare ^ against the new SACK entry 0. Now it takes the second branch,
cutting off the first part of SACK 0 so it matches the ^:

```
|...............000000.....11111......2222|
|---------------^
```

The rest of the branch, and the other iterations of the for loop now
will just change the offset to shift the remaining sacks to the start of
the receive buffer:

```
|000000.....11111......2222|
|^
```

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 replied:
Ah, thanks, I think I got it now.

vollkorn1982 commented on utcp.c:
```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+			if(offset + rxd > c->sacks[i].offset + c->sacks[i].len) { // merge
+				debug("Merge with end of SACK entry at %d\n", i);
+				c->sacks[i].len = offset + rxd - c->sacks[i].offset;
+				// TODO: handle potential merge with next entry
```
How crucial is this? Maybe add some debug output to find out under which circumstances and how often this occurs?

gsliepen replied:
On Thu, Oct 29, 2015 at 06:27:56AM -0700, vollkorn1982 wrote:

> > -       } else if(offset <= c->sacks[i].offset + c->sacks[i].len) {
> > -           if(offset + rxd > c->sacks[i].offset + c->sacks[i].len) { // merge
> > -               debug("Merge with end of SACK entry at %d\n", i);
> > -               c->sacks[i].len = offset + rxd - c->sacks[i].offset;
> > -               // TODO: handle potential merge with next entry
> 
> How crucial is this? Maybe add some debug output to find out under which circumstances and how often this occurs?

Nothing will break by not handling this (I tested that), but it is more
efficient to do it properly. It might occur when there is a lot of
packet reordering. I'll do this TODO.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 replied:
If it's just about efficiency let's do this later. This pull request is about getting it to work at all.

vollkorn1982 commented on utcp.c:
```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+			break;
+		} else if(offset < c->sacks[i].offset) {
+			if(offset + rxd < c->sacks[i].offset) { // insert before
+				if(!c->sacks[NSACKS - 1].len) { // only if room left
```
So if no room is left the received packet is silently dropped? How about at least a debug output?

gsliepen replied:
On Thu, Oct 29, 2015 at 06:29:40AM -0700, vollkorn1982 wrote:

> > -   // Packet loss or reordering occured. Store the data in the buffer.
> > -   ssize_t rxd = buffer_put_at(&c->rcvbuf, offset, data, len);
> >   +
> > -   // Make note of where we put it.
> > -   for(int i = 0; i < NSACKS; i++) {
> > -       if(!c->sacks[i].len) { // nothing to merge, add new entry
> > -           debug("New SACK entry %d\n", i);
> > -           c->sacks[i].offset = offset;
> > -           c->sacks[i].len = rxd;
> > -           break;
> > -       } else if(offset < c->sacks[i].offset) {
> > -           if(offset + rxd < c->sacks[i].offset) { // insert before
> > -               if(!c->sacks[NSACKS - 1].len) { // only if room left
> 
> So if no room is left the received packet is silently dropped?

Correct. Note that before this SACK mechanism, we would have silently
dropped any out-of-order packet.

> How about at least a debug output?

Sure.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 commented on utcp.c:
```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+
+static void handle_in_order(struct utcp_connection *c, const void *data, size_t len) {
+	// Check if we can process out-of-order data now.
+	if(c->sacks[0].len && len >= c->sacks[0].offset) {
```
Is it really never possible to have an empty `c->sacks` here and will `c->sacks[0]` always succeed?

gsliepen replied:
On Thu, Oct 29, 2015 at 06:37:05AM -0700, vollkorn1982 wrote:

> > +static void handle_in_order(struct utcp_connection *c, const void *data, size_t len) {
> > -   // Check if we can process out-of-order data now.
> > -   if(c->sacks[0].len && len >= c->sacks[0].offset) {
> 
> Is it really never possible to have an empty `c->sacks` here and will `c->scacks[0]` always succeed?

Yes. From utcp_priv.h:

struct utcp_connection {
    // ...
    struct sack sacks[NSACKS];
    // ...
};

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 commented on utcp.c:
```diff
@@ -635,63 +893,195 @@ ssize_t utcp_recv(struct utcp *utcp, const void *data, size_t len) {
+					struct timeval now, diff;
+					gettimeofday(&now, NULL);
+					timersub(&now, &c->rtt_start, &diff);
+					update_rtt(c, diff.tv_sec * 1000000 + diff.tv_usec);
```
I have seen this `1000000` several times now. Maybe it's better to make this one a `#define` if we ever need to change this.

gsliepen replied:
On Thu, Oct 29, 2015 at 06:47:27AM -0700, vollkorn1982 wrote:

> > -                   update_rtt(c, diff.tv_sec \* 1000000 + diff.tv_usec);
> 
> I have seen this `1000000` several times now. Maybe it's better to make this one a `#define` if we ever need to change this.

#define USEC_PER_SEC 1000000

It would only ever need to change when the SI definition of "micro" changes ;)

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 replied:
I was thinking more in "oh, let's triple those times for some obscure reason". Also writing 1000000 in code often makes you vulnerable for miscounting 0s. Like once you'll enter 100000 and will have a very long and annoying debugging session.

vollkorn1982 commented on utcp.c:
```diff
@@ -1131,33 +1442,31 @@ struct timeval utcp_timeout(struct utcp *utcp) {
+					c->recv(c, NULL, 0);
+				continue;
+			}
+			if(timercmp(&c->conn_timeout, &next, <))
```
Hm. `timercmp` is posix only as well, isn't it?

gsliepen replied:
On Thu, Oct 29, 2015 at 07:41:12AM -0700, vollkorn1982 wrote:

> > -           if(timercmp(&c->conn_timeout, &next, <))
> 
> Hm. `timercmp` is posix only as well, isn't it?

Good point. Actually, according to the manpage, it's not even POSIX, but
rather a BSD thing.

I should rewrite everything to use clock_gettime(), since that is the
POSIXest way to do time nowadays. Also, clock_gettime(CLOCK_MONOTONIC)
is safer to use, and perhaps even CLOCK_MONOTONIC_COARSE can be used in
UTCP.

There is no standard C99 function to get time with a resolution better
than a second, so for Windows we need to write our own clock_gettime(),
like suggested here:

http://stackoverflow.com/questions/5404277/porting-clock-gettime-to-windows

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

vollkorn1982 replied:
Well, I think we can postpone this rewrite for a later time when we want to take care of the windows port. The current goal is to get ack'ing and retransmit to work.

vollkorn1982 commented on utcp.c:
```diff
@@ -1177,21 +1486,22 @@ bool utcp_is_active(struct utcp *utcp) {
-	utcp->timeout = 60;
+	utcp->mtu = DEFAULT_MTU;
+	utcp->timeout = DEFAULT_USER_TIMEOUT; // s
+	utcp->rto = START_RTO; // us
```
What is meant by `s` and `us`?

vollkorn1982 replied:
Ah, signed and unsigned?

gsliepen replied:
On Thu, Oct 29, 2015 at 07:42:17AM -0700, vollkorn1982 wrote:

> > -   utcp->timeout = DEFAULT_USER_TIMEOUT; // s
> > -   utcp->rto = START_RTO; // us
> 
> What is meant by `s` and `us`?

Seconds and microseconds. I'll expand it.

## 

Met vriendelijke groet / with kind regards,
     Guus Sliepen guus@tinc-vpn.org

Commit:
Always check the return value of malloc().

Commit:
Don't use ?: without the middle argument.

Commit:
Make max() an inline function.

Commit:
Clarify description of sack_consume().

Commit:
Add debug message when dropping a packet because all SACK entries are used.

Commit:
Define USEC_PER_SEC, use "sec" and "usec" in comments.

everHannes committed:
fix undef mingw max macro

everHannes committed:
fix print_packet data offset and separate and comment binToHex conversion
